[hma] Add a settings pane for getting index last modified and rebuilding on demand

schatten opened the pull request:
Summary
---------
With banks, indexing runs on a pre-determined frequency. This may not be fast enough for our users when they are demoing something or need to test changes out.

This shows the last_modified times of the current set of indexes on the 'settings' portion of the HMA UI and allows a user to trigger an async rebuild of the index.

Test Plan
---------
mypy, black, py.test on local.

**Screenshot 1: the new UI pane with the rebuild button**
<img width="499" alt="Screen Shot 2021-12-13 at 22 37 58" src="https://user-images.githubusercontent.com/217056/145929222-85d317bb-b210-424b-88a6-ff7eb71e97d5.png">

**Screenshot 2: logs of indexer when the rebuild button was clicked**
<img width="1114" alt="Screen Shot 2021-12-13 at 22 38 10" src="https://user-images.githubusercontent.com/217056/145929253-5e9eddf7-aa6b-4683-8d6d-b674d6e19971.png">

BarrettOlson commented on hasher-matcher-actioner/hmalib/indexers/s3_indexers.py:
```diff
@@ -35,6 +36,8 @@ class S3BackedInstrumentedIndexMixin:
     Also overrides methods to wrap instrumentation with hmalib.metrics library.
     """
 
+    INDEXES_PREFIX = "index/"
```
(nonblocking) We haven't been consistent about this being a variable vs just locally defining it. It used to be something defined in terraform...

BarrettOlson replied:
nevermind, looks like this is a step in the right direction regardless

BarrettOlson commented on hasher-matcher-actioner/hmalib/indexers/s3_indexers.py:
```diff
@@ -54,6 +57,16 @@ def load(cls, bucket_name: str):
+        objects = get_s3_client().list_objects_v2(
+            Bucket=bucket_name, Prefix=cls.INDEXES_PREFIX
+        )
+        return max(map(lambda item: item["LastModified"], objects["Contents"]))
```
How is "LastModified" stored? depending on the timestamp counting up wouldn't this actually be the most recently modified?

schatten replied:
Yes. This is the most recently modified timestamp.

I am not accounting for a case where one index is built but the other is not.

BarrettOlson commented on hasher-matcher-actioner/hmalib/lambdas/api/indexes.py:
```diff
@@ -0,0 +1,59 @@
+    def all_indexes_last_modified() -> IndexesLastModifiedResponse:
+        """
+        Returns the max of last_modified time of all indexes. Read as: when was
+        the latest index rebuilt.
```
hmm, 'latest' or 'oldest' am I miss understanding these as contradictory.

BarrettOlson reviewed (approved):
Approving to unblock. Thanks
two asides though
1) can you confirm intended behavior is correct e.g. "oldest" vs "latest"

2) to me this index button is really nice to have but it could just be a button/card on a existing settings or bank page instead of its own dedicated one.

schatten commented:
>     1. can you confirm intended behavior is correct e.g. "oldest" vs "latest"

Fixed. It is latest. Not oldest. I started with oldest but quickly realized this would be really hard to get right. So switched everywhere but the method name. This is now addressed.

>     2. to me this index button is really nice to have but it could just be a button/card on a existing settings or bank page instead of its own dedicated one.

I agree. I struggled for a bit, but I could not figure out a place where this would fit better. Maybe a global settings page, but we don't have one. The equivalent settings for actions and action rules have their own page, so decided to go down this path, but am not beholden to it. 

Happy to discuss and move elsewhere once this lands.
